### DISCUSSION TITLE
wez with agent forwarding

### DISCUSSION THREAD

**Discussion by flavius** [wezterm/wezterm#988] [Q&A]
So I have the agent started, key loaded, `ssh-add -L` shows it.

Now I want to connect to a remote wez ssh host and keep my agent around.

with other emulators it works and I think I could make it work also with tmux upon rea-attaching.

I also have `ForwardAgent yes` in my `.ssh/config` and logging in to the server and doing `ssh-add -L` shows the key also.

The issue is then How can I get forwarding to work with ssh multiplexing/domains?

My current config:

```
local wezterm = require 'wezterm';

return  {
  disable_default_key_bindings = true,

  font_size = 10,

  leader = {key="a", mods="CTRL", timeout_milliseconds=1000},
  keys = {
    { key="L", mods="LEADER", action="ShowLauncher"},
    { key="c", mods="LEADER", action=wezterm.action{SpawnTab="CurrentPaneDomain"}},
    { key="t", mods="LEADER", action=wezterm.action{SpawnTab={DomainName="controller"}}},
    { key="u", mods="LEADER", action=wezterm.action{SpawnCommandInNewTab={
      -- args={"sudo", "bash", "--rcfile", "development/bin/activate", "-r"},
      args={"bash"}, -- HERE I need the agent to be forwarded
      domain={DomainName="controller"},
      cwd="/home/flaasp/unnamed/unnamed-master",
      set_environment_variables = {
        TEST = "a value"
      },
    }}},
  },
    ssh_domains = {
    {
      -- This name identifies the domain
      name = "controller",
      -- The address to connect to
      remote_address = "controller",
      -- The username to use on the remote host
      username = "flaasp",
    },
    unix_domains = {
      {
        name = "tunnels",
        connect_automatically = false,
      },
    },
  }
}

```

I marked the relevant line above with `HERE I need the agent to be forwarded`

In the code, I could find:

https://github.com/wez/wezterm/blob/c22765c1a76aea8c7cac6eac99a549962058af10/wezterm-mux-server/src/main.rs#L130-L146

https://github.com/wez/wezterm/blob/ba42367fbf6e8a8a48adb2fc3869dc625b18a8be/wezterm-ssh/src/config.rs#L524-L528

Which makes me believe it's not possible?

**Comment by wez** [comment]
libssh2 doesn't properly implement support for agent forwarding:

https://github.com/wez/wezterm/commit/c6cd1a6a33fa9850c920ae78f5f7e8bd99e55ac5

Until it does, wezterm is also unable to provide agent forwarding.

**Reply by barisozdag** [reply]
I think that time has come: https://github.com/libssh2/libssh2/issues/535#issuecomment-1536804881

**Comment by carlesso** [comment]
Poking at this, the libssh2 issue seems addressed, is this unblocked now?

**Comment by amsesk** [comment]
I am also very interested in this. Lack of agent forwarding is the only thing keeping me on kitty/tmux for now.

**Comment by wez** [comment]
as of 4af418fddd0ed2d9a8861007112006fc657ecbac, the mux server now automatically maintains SSH_AUTH_SOCK based on the most recently active multiplexer client. There have been additional commits to enable agent forwarding, if your ssh config enables it, for ssh domains.

**Reply by rgwood** [reply]
FYI this commit seems to have broken the use of `ssh-agent` on Windows (I noticed on latest `main` and bisected).

Repro steps:

1. Set `ssh_backend = "Ssh2` to get the agent working as per [this discussion](https://github.com/wez/wezterm/discussions/3772#discussioncomment-7201688)
2. Try running `ssh-add`, get a "Error connecting to agent: No such file or directory" error

I've noticed that my `SSH_AUTH_SOCK` env var is set to something like `C:\Users\reill\.local/share/wezterm\agent.3568` after that commit, but unsetting it doesn't fix the issue.

It's late here but I can create an issue tomorrow if that would be helpful.

edit: after reading through the commit, I was able to set `mux_enable_ssh_agent = false` to fix this

**Reply by Jaykul** [reply]
Yeah, I agree, I just recently switched to nightly wezterm, and this broke my `earthly` builds. I started getting:
> Error: ssh agent provider: CreateFile C:\Users\Jaykul\.local/share/wezterm\agent.30328: The file cannot be accessed by the system.

I normally use KeeAgent in KeePass for my ssh agent, but I have it configured to enable the agent for Windows' OpenSSH and create a WSL-compatible socket file (which has been working fine with WSL2)....
